source/common/formatter/stream_info_formatter.cc

```diff
@@ -180,42 +197,66 @@ FilterStateFormatter::filterState(const StreamInfo::StreamInfo& stream_info) con
   return nullptr;
 }
 
+struct StringFieldVisitor {
+  absl::optional<std::string> operator()(int64_t val) { return absl::StrCat(val); }
+  absl::optional<std::string> operator()(absl::string_view val) { return std::string(val); }
+  absl::optional<std::string> operator()(absl::monostate) { return {}; }
+};
+
 absl::optional<std::string>
 FilterStateFormatter::format(const StreamInfo::StreamInfo& stream_info) const {
   const Envoy::StreamInfo::FilterState::Object* state = filterState(stream_info);
   if (!state) {
     return absl::nullopt;
   }
 
-  if (serialize_as_string_) {
+  switch (format_) {
+  case FilterStateFormat::String: {
     absl::optional<std::string> plain_value = state->serializeAsString();
     if (plain_value.has_value()) {
       SubstitutionFormatUtils::truncate(plain_value.value(), max_length_);
       return plain_value.value();
     }
     return absl::nullopt;
   }
-
-  ProtobufTypes::MessagePtr proto = state->serializeAsProto();
-  if (proto == nullptr) {
-    return absl::nullopt;
-  }
+  case FilterStateFormat::Proto: {
+    ProtobufTypes::MessagePtr proto = state->serializeAsProto();
+    if (proto == nullptr) {
+      return absl::nullopt;
+    }
 
 #ifdef ENVOY_ENABLE_FULL_PROTOS
-  std::string value;
-  const auto status = Protobuf::util::MessageToJsonString(*proto, &value);
-  if (!status.ok()) {
-    // If the message contains an unknown Any (from WASM or Lua), MessageToJsonString will fail.
-    // TODO(lizan): add support of unknown Any.
-    return absl::nullopt;
-  }
+    std::string value;
+    const auto status = Protobuf::util::MessageToJsonString(*proto, &value);
+    if (!status.ok()) {
+      // If the message contains an unknown Any (from WASM or Lua), MessageToJsonString will fail.
+      // TODO(lizan): add support of unknown Any.
+      return absl::nullopt;
+    }
 
-  SubstitutionFormatUtils::truncate(value, max_length_);
-  return value;
+    SubstitutionFormatUtils::truncate(value, max_length_);
+    return value;
 #else
-  PANIC("FilterStateFormatter::format requires full proto support");
-  return absl::nullopt;
+    PANIC("FilterStateFormatter::format requires full proto support");
+    return absl::nullopt;
 #endif
+  }
+  case FilterStateFormat::Field:
+    if (factory_) {
+      const auto reflection = factory_->reflect(state);
+      if (reflection) {
```

Comment:
nit: consider breaking up nesting by returning nullopt if conditions fail for further processing

Reply:
Done.